### DISCUSSION TITLE
How to send data to the property inspector?

### DISCUSSION THREAD

**Discussion by MaterialFoundry** [elgatosf/elgato#141] [Help] [answered]
How do I send data from the plugin to the property inspector using the Node.js tools?
With the old template I could do something like:
```js
$SD.connection.sendJSON({
    event: "sendToPropertyInspector",
    action,
    context,
    payload
});
```
What is the equivalent for the Node.js tools?

I was hoping for something like `streamDeck.actions.sendToPropertyInspector(data)`, but I can't find anything like that.
The only relevant thing I can find is [this](https://docs.elgato.com/streamdeck/sdk/references/websocket/plugin#sendtopropertyinspector), but I can't find any info on how to send anything over the WebSocket connection.

**Comment by MaterialFoundry** [comment, marked as answer]
I figured it out, for anyone else struggling with this:

To send data to the PI: 
```js
streamDeck.ui.current?.sendToPropertyInspector(payload);
```
Which, in the PI code can be received using: 
```js
SDPIComponents.streamDeckClient.sendToPropertyInspector.subscribe((ev) => {
   //do stuff here
})`
```
The other way around, to send data to the plugin:
```js
SDPIComponents.streamDeckClient.send('sendToPlugin', payload);
```
Which can be received the SingletonAction class using:
```js
override async onSendToPlugin(ev: SendToPluginEvent): Promise<void> {
  //do stuff here
}
```

**Comment by willievaldez** [comment]
Related to this, I'm wondering if anyone can give me some insight on the ["Data Source"](https://sdpi-components.dev/docs/helpers/data-source) PI feature. It says I can add a datasource field to my sdpi html element, which Ive done like this (trying to directly copy the documentation):
```
        <sdpi-select setting="light_1.color" placeholder="-NONE-" datasource="getColors"
        loading="Fetching colors...">
        </sdpi-select>
```
and then I listen to the get request in the plugin via `onSendToPlugin`, and then I return my data via `sendToPropertyInspector`:
``` 
	override async onSendToPlugin(ev: SendToPluginEvent<JsonObject, LightGroupSettings>): Promise<void> {
		if (ev.payload.event == "getColors")
		{
			streamDeck.ui.current?.sendToPropertyInspector({
				action: ev.payload.action,
				context: ev.payload.context,
				event: ev.payload.event,
				payload: {
					event: "getColors",
					items: [{
						label: 'Black',
						value: '3'
					}, {
						label: 'White',
						value: '4'
					}]
				}
			})
		}
		streamDeck.connect();
	}
```
The dropdown stops displaying "Fetching colors...", so it seems to be receiving a message from the plugin, but when clicking on it, there are no selectable options. Am I missing any steps here? Why don't I see "Black" and "White" as dropdown options?

**Comment by willievaldez** [comment]
Update: it looks like the API doesnt match [the documentation](https://sdpi-components.dev/docs/helpers/data-source#plugin). My `sendToPropertyInspector` function works properly when I just set the event and items:
```
	override async onSendToPlugin(ev: SendToPluginEvent<JsonObject, LightGroupSettings>): Promise<void> {
		if (ev.payload.event == "getColors")
		{
			streamDeck.ui.current?.sendToPropertyInspector({
				event: "getColors",
				items: [{
					label: 'Black',
					value: '3'
				}, {
					label: 'White',
					value: '4'
				}]
			})
		}
		streamDeck.connect();
	}
```

**Reply by Elephant42** [reply]
Many thanks for sharing your info - I had exactly the same problem trying to return a list of fonts - was tearing my hair out until I stumbled across your post.  I wonder what the quickest way is to get the docs updated - they are still incorrect as of the date of this post.
